feat: Build plots and charts from tables, with Python code generation

Light2Dark opened the pull request:
## 📝 Summary

<!--
Provide a concise summary of what this pull request is addressing.

If this PR fixes any issues, list them here by number (e.g., Fixes #123).
-->
https://github.com/user-attachments/assets/fc57c39c-565f-4433-b586-7052b52941b5

- Release chart builder in tables. 
- Introduce some QOL improvements in charts (better specification defaults, tooltips, collapsible sidebar) and some bug fixes.
- Table panels (bottom left) now conditionally appear depending on the mode marimo is run in.
- In edit mode, all panels are available while in other modes only the row-viewer panel appears.

## 🔍 Description of Changes

<!--
Detail the specific changes made in this pull request. Explain the problem addressed and how it was resolved. If applicable, provide before and after comparisons, screenshots, or any relevant details to help reviewers understand the changes easily.
-->

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://marimo.io/discord?ref=pr), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [x] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

<!--
Tag potential reviewers from the community or maintainers who might be interested in reviewing this pull request.

Your PR will be reviewed more quickly if you can figure out the right person to tag with @ -->

akshayka reviewed (changes requested):
Not necessarily changes to the code — but since this PR GAs charts, we should definitely highlight it in release notes and feature this PR prominently there.

I've added the release highlight tag to the PR.

Can you:
* Update the PR description to describe the entire charts feature in a few paragraphs.
* Add media, such as a video, that demos the feature?

akshayka reviewed (commented):
Some comments from playing with charts.

* Can we use `ChartSpline` instead of `ChartBar` for the icon in the table footer?
* Bar chart aggregation dropdown menu. I feel like `Bin` is going to be the most common aggregation. Can we move it to the top?
* Faceting section. Can you add a `?` icon with a tooltip that explains what faceting is?
* Faceting. Does it really make sense to facet by numerical columns? I feel like only categorical columns should support faceting.
* More generally, has much thought gone into eliminating values from various chart fields when they don't make sense? For example, for a line chart, the only y-value that makes sense to me is numerical.
* We should tell the user somehow that charts won't be persisted. It would be a sad first impression if a user carefully crafted lots of charts but didn't save them to their notebook (not knowing they wouldn't be persisted) and lost them all.
* Scatter. Sorting the x-axis by ascending/descending doesn't appear to do anything (my axes are both numerical). Sorting doesn't make much sense to me as an operation for a scatter chart anyway. It also doesn't seem to do anything on Heatmap. Can we just remove the axis sorting?
* Scatter color-by stack doesn't make sense to me. For example here the y-axis is wrong (the max displacement in this plot is ~350, and yet when I stack the values go up to 8000)

Releasing this feature is a big (awesome!) change. We should also have a docs page under > Working with data (either a new page, or a subsection in  "Plotting"). The docs can be in a separate PR.

Light2Dark commented:
Thanks for testing @akshayka , I've made those changes in https://github.com/marimo-team/marimo/pull/5392 and going to mark this as draft until those are merged.

Generally, we can include those fields (eg. faceting numerical and categorical y-axis), but we need to set better defaults. Numerical faceting should be binned for example.

manzt reviewed (approved):
Looks good to me 👍 I was initially going to suggest wrapping these settings in a `React.Context`, but I noticed the rest of the props are passed down in a similar way and would prefer to be consistent.

Light2Dark commented:
- changed to chartspline
- bin moved to the top
- added binning to numerical faceting by default
- scatter charts have grid on by default
- removed sorting for scatter, heatmap charts
- stacking only allowed for bar and line charts
- for categorical fields on y-axis, a default aggregation is set to count
- added a tooltip to provide persistence info

![CleanShot 2025-07-04 at 00 52 39](https://github.com/user-attachments/assets/edb88a37-ef63-40f7-9736-6348201229ef)

akshayka reviewed (approved):
I pushed two small things:

* Updated the tooltip.
* Don't show the add code button when there's no code to add.

I also updated the PR title to help with release notes.
